drv/fpga-server/src/main.rs

```diff
@@ -0,0 +1,272 @@
+        Ok(self.device.device_state()?)
+    }
+
+    fn device_id(&mut self, _: &RecvMessage) -> Result<u32, RequestError> {
+        Ok(self.device.device_id()?)
+    }
+
+    fn application_enabled(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<u8, RequestError> {
+        Ok(self.device.application_enabled()? as u8)
+    }
+
+    fn set_application_enable(
+        &mut self,
+        _: &RecvMessage,
+        enabled: u8,
+    ) -> Result<(), RequestError> {
+        Ok(self.device.set_application_enable(enabled != 0)?)
+    }
+
+    fn reset_application(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<(), RequestError> {
+        Ok(self
+            .device
+            .reset_application(self.application_reset_ticks)?)
+    }
+
+    fn start_bitstream_load(
+        &mut self,
+        _: &RecvMessage,
+        bitstream_type: BitstreamType,
+    ) -> Result<(), RequestError> {
+        if let BitstreamType::Compressed = bitstream_type {
+            self.decompressor = Some(gnarle::Decompressor::default())
+        }
+        self.device.start_bitstream_load()?;
+        ringbuf_entry!(Trace::StartBitstreamLoad(bitstream_type));
+        Ok(())
+    }
+
+    fn continue_bitstream_load(
+        &mut self,
+        _: &RecvMessage,
+        data: LenLimit<Leased<R, [u8]>, 128>,
+    ) -> Result<(), RequestError> {
+        data.read_range(0..data.len(), &mut self.buffer[..data.len()])
+            .map_err(|_| RequestError::Fail(ClientError::WentAway))?;
+
+        let chunk = &mut &self.buffer[..data.len()];
+        let mut decompress_buffer = [0; 256];
+
+        match self.decompressor.as_mut() {
+            Some(decompressor) => {
+                while !chunk.is_empty() {
+                    let decompressed_chunk = gnarle::decompress(
+                        decompressor,
+                        chunk,
+                        &mut decompress_buffer,
+                    );
+                    self.device.continue_bitstream_load(decompressed_chunk)?;
+                }
+            }
+            None => self.device.continue_bitstream_load(chunk)?,
+        }
+
+        ringbuf_entry!(Trace::ContinueBitstreamLoad(data.len()));
+        Ok(())
+    }
+
+    fn finish_bitstream_load(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<(), RequestError> {
+        self.decompressor = None;
+        self.device
+            .finish_bitstream_load(self.application_reset_ticks)?;
+        ringbuf_entry!(Trace::FinishBitstreamLoad);
+        Ok(())
+    }
+
+    fn application_read(
+        &mut self,
+        _: &userlib::RecvMessage,
+        addr: u16,
+        data: WriteDataLease,
+    ) -> Result<(), RequestError> {
+        let header = ApplicationRequestHeader {
+            cmd: 0x1,
+            addr: U16::new(addr),
+        };
+
+        self.application
+            .lock(spi_api::CsState::Asserted)
+            .map_err(FpgaError::from)?;
+        self.application
```

Comment:
If these functions fail, do we still need to call `self.application.release()`?

Reply:
This has been resolved using `Drop` types.